Updates to README (#153)
klaviyo/klaviyo-swift-sdk: README.md

# klaviyo-swift-sdk

![CI status](https://github.com/klaviyo/klaviyo-swift-sdk/actions/workflows/swift.yml/badge.svg)
[![Swift Package Manager](https://img.shields.io/badge/Swift_Package_Manager-compatible-orange?style=flat-square)](https://img.shields.io/badge/Swift_Package_Manager-compatible-orange?style=flat-square)
![SPM version](https://img.shields.io/github/v/release/klaviyo/klaviyo-swift-sdk)
[![Version](https://img.shields.io/cocoapods/v/KlaviyoSwift.svg?style=flat)](http://cocoapods.org/pods/KlaviyoSwift)
[![License](https://img.shields.io/cocoapods/l/KlaviyoSwift.svg?style=flat)](http://cocoapods.org/pods/KlaviyoSwift)
![Minimum deployment version](https://img.shields.io/badge/minimum_iOS_deployment_target-iOS13-brightgreen)

## Contents

> - [Overview](#overview)

Reviewer: ```suggestion
```

- [Installation](#installation)
- [Initialization](#initialization)
- [Profile Identification](#profile-identification)
  - [Reset Profile](#reset-profile)
  - [Anonymous Tracking Notice](#anonymous-tracking-notice)
- [Event tracking](#event-tracking)
- [Push Notifications](#push-notifications)
  - [Prerequisites](#prerequisites)
  - [Setup](#setup)
  - [Collecting Push Tokens](#collecting-push-tokens)
    - [Push Tokens And Multiple Profiles](#push-tokens-and-multiple-profiles)
  - [Request Push Notification Permission](#request-push-notification-permission)
  - [Receiving Push Notifications](#receiving-push-notifications)
    - [Tracking Open Events](#tracking-open-events)
    - [Rich Push](#rich-push)
    - [Deep Linking](#deep-linking)
      - [Option 1: URL Schemes](#option-1-URL-schemes)
      - [Option 2: Universal Links](#option-2-universal-links)
- [Additional Details](#additional-details)
  - [Sandbox Support](#sandbox-support)
  - [SDK Data Transfer](#sdk-data-transfer)
  - [Retries](#retries)
  - [License](#license)

## Overview

The Klaviyo Swift SDK allows developers to incorporate Klaviyo's analytics and push notification functionality into their iOS applications.

> The SDK assists in identifying users and tracking events via [Klaviyo client APIs](https://developers.klaviyo.com/en/reference/api_overview).

Reviewer: ```suggestion
The SDK assists in identifying users and tracking events via [Klaviyo Client APIs](https://developers.klaviyo.com/en/reference/api_overview).
```

# klaviyo-swift-sdk

![CI status](https://github.com/klaviyo/klaviyo-swift-sdk/actions/workflows/swift.yml/badge.svg)
[![Swift Package Manager](https://img.shields.io/badge/Swift_Package_Manager-compatible-orange?style=flat-square)](https://img.shields.io/badge/Swift_Package_Manager-compatible-orange?style=flat-square)
![SPM version](https://img.shields.io/github/v/release/klaviyo/klaviyo-swift-sdk)
[![Version](https://img.shields.io/cocoapods/v/KlaviyoSwift.svg?style=flat)](http://cocoapods.org/pods/KlaviyoSwift)
[![License](https://img.shields.io/cocoapods/l/KlaviyoSwift.svg?style=flat)](http://cocoapods.org/pods/KlaviyoSwift)
![Minimum deployment version](https://img.shields.io/badge/minimum_iOS_deployment_target-iOS13-brightgreen)

## Contents
- [Overview](#overview)
- [Installation](#installation)
- [Initialization](#initialization)
- [Profile Identification](#profile-identification)
  - [Reset Profile](#reset-profile)
  - [Anonymous Tracking Notice](#anonymous-tracking-notice)
- [Event tracking](#event-tracking)
- [Push Notifications](#push-notifications)
  - [Prerequisites](#prerequisites)
  - [Setup](#setup)

>   - [Collecting Push Token](#collecting-push-token)

Reviewer 2: Tokens?

Reviewer 2: Maybe singular is ok though in this case!

Reviewer 3: I would make it plural!

Author: Made it plural.

    - [Push Tokens And Multiple Profiles](#push-tokens-and-multiple-profiles)
  - [Request Push Notification Permission](#request-push-notification-permission)
  - [Receiving Push Notifications](#receiving-push-notifications)

>     - [Tracking opens](#tracking-opens)

Reviewer: ```suggestion
    - [Tracking Open Events](#Tracking-Open-Events)
```

>     - [Rich Push Notifications](#rich-push-notifications)

Reviewer 3: In the Android readme, it looks like Evan has rich push before tracking opens. Is there a reason you have the order flipped? Should we be consistent?

Author: Yeah I should have called this out - I did opens first since that seems more important/required than rich push and also on iOS rich push is additional setup where as on android it's mostly taken care on our SDK. Let me know what you think?

Reviewer: My order was more like chronological
Rich Push (displaying) -> Open -> Deep Links

Reviewer: For my part, I don't mind the variation in order if the header names are the same.

Reviewer 3: If that order makes the most sense for iOS, let's go with it @ajaysubra .

Author: switched it up. thanks!

>     - [Rich Push Notifications](#rich-push-notifications)

Reviewer: ```suggestion
    - [Rich Push](#Rich-Push)
```

    - [Deep Linking](#deep-linking)
      - [Option 1: URI Schemes](#option-1-uri-schemes)
      - [Option 2: Universal Links](#option-2-universal-links)
- [Additional Details](#additional-details)
  - [Sandbox Support](#sandbox-support)
  - [SDK Data Transfer](#sdk-data-transfer)
  - [Retries](#retries)
  - [License](#license)
  - [UserDefaults access (SDK version < 3.0.0)](#userdefaults-access-sdk-version--300)

## Overview

> The Klaviyo Swift SDK allows developers to incorporate Klaviyo's analytics and push notification functionality in their iOS applications.

Reviewer 4: nit: `The Klaviyo Swift SDK allows developers to incorporate Klaviyo's analytics and push notification functionality into their iOS applications.` 

`into their` instead of `in their` maybe?

The SDK assists in identifying users and tracking events via [Klaviyo client APIs](https://developers.klaviyo.com/en/reference/api_overview).
To reduce performance overhead, API requests are queued and sent in batches.
The queue is persisted to local storage so that data is not lost if the device is offline or the app is terminated.

Once integrated, your marketing team will be able to better understand your app users' needs and send them timely messages via APNs.

## Installation

<details>
<summary>Swift Package Manager [Recommended]</summary>

KlaviyoSwift is available via [Swift Package Manager](https://swift.org/package-manager). Follow the steps below to install.

1. Open your project and navigate to your project’s settings.
2. Select the **Package Dependencies** tab and click on the **add** button below the packages list.

> 3. Enter the URL of the Swift SDK repository `https://github.com/klaviyo/klaviyo-swift-sdk` in the text field, and you should see the result.

Author: 🤦🏽 yeah I agree this does sound right. Essentially, I'm trying to point out that you should see the SDK show up as a results on this screen like below. 
![image](https://github.com/klaviyo/klaviyo-swift-sdk/assets/118314354/f79bcf83-d047-4076-b144-04b1c6e206e0)

I'll update this to be more clear.

> 3. Enter the URL of the Swift SDK repository `https://github.com/klaviyo/klaviyo-swift-sdk` in the text field, and you should see the result.

Author: >3. Enter the URL of the Swift SDK repository `https://github.com/klaviyo/klaviyo-swift-sdk` in the text field. This should bring up the package on the screen.

Does this sound better?

The following code demonstrates how to set profile identifiers:

```swift
// organization, title, image, location and additional properties (dictionary) can also be set using the below constructor
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// or setting individual properties
KlaviyoSDK().set(profileAttribute: .firstName, value: "Blob")
KlaviyoSDK().set(profileAttribute: .lastName, value: "Jr.")
```

### Reset Profile
To start a new profile altogether (e.g. if a user logs out) either call `KlaviyoSDK().resetProfile()` to clear the currently tracked profile identifiers,
or use `KlaviyoSDK().set(profile: profile)` to overwrite it with a new profile object.

```swift
// start a profile for Blob Jr.
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// stop tracking Blob Jr.
KlaviyoSDK().resetProfile()

// start a profile for Robin Hood
let profile = Profile(email: "robin@hood.com",  firstName: "Robin",  lastName: "Hood")
KlaviyoSDK().set(profile: profile)
```
### Anonymous Tracking Notice

> By default, Klaviyo will begin tracking unidentified users in your app when a push token is set, or if it receives an event to track.
> This means you will be able to track events or set a push notification token from users in your app without any user information provided.
> When an email or other primary identifier is provided, Klaviyo will merge the data from the anonymous user to a new identified user.

Reviewer: ```suggestion
Klaviyo will track unidentified users with an autogenerated ID whenever a push token is set or an event is created. 
That way, you can collect push tokens and track events prior to collecting profile identifiers such as email or phone number.
When an identifier is provided, Klaviyo will merge the anonymous user with an identified user.
```

Reviewer: workshopped with mark on the android readme

Author: Thanks! Commited your suggestion.

for your Klaviyo account, also known as your Site ID.

```swift
// AppDelegate

import KlaviyoSwift

class AppDelegate: UIResponder, UIApplicationDelegate {

    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        KlaviyoSDK().initialize(with: "YOUR_KLAVIYO_PUBLIC_API_KEY")
        return true
    }
}
```

The SDK **should** be initialized before any other Klaviyo SDK methods are called.

## Profile Identification
The SDK provides methods to identify your users as Klaviyo profiles via the [Create Client Profile API](https://developers.klaviyo.com/en/reference/create_client_profile).
A profile can be identified by any combination of the following:

* External ID: A unique identifier used by customers to associate Klaviyo profiles with profiles in an external system, such as a point-of-sale system. Format varies based on the external system.
* Individual's email address
* Individual's phone number in [E.164 format](https://help.klaviyo.com/hc/en-us/articles/360046055671#h_01HE5ZYJEAHZKY6WZW7BAD36BG)

> These above Identifiers are persisted to local storage so that the SDK can keep track of the current user/profile for you when you make event requests or want to set a push token etc.

Reviewer: ```suggestion
These above identifiers are persisted to local storage so that the SDK can keep track of the current user/profile for you when you make event requests or want to set a push token etc.
```

Profile identifiers can be set all at once or individually. Either way, the SDK will group and batch API calls to improve performance.

The following code demonstrates how to set profile identifiers:

```swift
// organization, title, image, location and additional properties (dictionary) can also be set using the below constructor
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// or setting individual properties
KlaviyoSDK().set(profileAttribute: .firstName, value: "Blob")
KlaviyoSDK().set(profileAttribute: .lastName, value: "Jr.")
```

### Reset Profile
To start a new profile altogether (e.g. if a user logs out) either call `KlaviyoSDK().resetProfile()` to clear the currently tracked profile identifiers,
or use `KlaviyoSDK().set(profile: profile)` to overwrite it with a new profile object.

```swift
// start a profile for Blob Jr.
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// stop tracking Blob Jr.
KlaviyoSDK().resetProfile()

// start a profile for Robin Hood

> let profile = Profile(email: "robin@hood.com",  firstName: "Robin",  lastName: "Hood")

Reviewer: 🏹

KlaviyoSDK().set(profile: profile)
```

> ### Anonymous Tracking Notice

Reviewer: A good note that we should copy over to the other two readmes

Author: Thanks!

By default, Klaviyo will begin tracking unidentified users in your app when a push token is set, or if it receives an event to track.
This means you will be able to track events or set a push notification token from users in your app without any user information provided.
When an email or other primary identifier is provided, Klaviyo will merge the data from the anonymous user to a new identified user.

> Prior to version 1.7.0, the Klaviyo SDK used the [Apple identifier for vendor (IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) to facilitate anonymous tracking.

Reviewer: Wonder at what point we can remove this, we're on 3.0 now, referring back to 1.7 seems ooold

Author: Yeah you are right. I removed this - @ndurell let me know if that's okay?
CC: @mark-piana

## Push Notifications

### Prerequisites

* An apple developer [account](https://developer.apple.com/).
* Configure [iOS push notifications](https://help.klaviyo.com/hc/en-us/articles/360023213971) in Klaviyo account settings.

### Setup

* Enable push notification capabilities in your Xcode project. The section "Enable the push notification capability" in this [Apple developer guide](https://developer.apple.com/documentation/usernotifications/registering_your_app_with_apns#2980170) provides detailed instructions.
* If you intend to use rich push notifications add a [Notification service extension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension) to your xcode project.
A notification service app extension ships as a separate bundle inside your iOS app. To add this extension to your app:
  * Select File > New > Target in Xcode.
  * Select the Notification Service Extension target from the iOS > Application extension section.
  * Click Next.
  * Specify a name and other configuration details for your app extension.
  * Click Finish.

> > ⚠️ By default the deployment target of your notification service extension might be the latest iOS version and not

Reviewer 2: Maybe Put a note about adjust the deployment target to match the app.

Author: Done.

to request a push token from APNs. This is typically done in the [`application:didFinishLaunchingWithOptions:`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application) method of your app delegate.
* Implement the delegate method [`application:didRegisterForRemoteNotificationsWithDeviceToken`](https://developer.apple.com/documentation/appkit/nsapplicationdelegate/1428766-application)
in your application delegate to receive the push token from APNs and register it with Klaviyo.

Below is the code to do both of the above steps:
```swift
import KlaviyoSwift

func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
    KlaviyoSDK().initialize(with: "YOUR_KLAVIYO_PUBLIC_API_KEY")

    UIApplication.shared.registerForRemoteNotifications()

    return true
}

func application(_ application: UIApplication, didRegisterForRemoteNotificationsWithDeviceToken deviceToken: Data) {
    KlaviyoSDK().set(pushToken: deviceToken)
}
```
#### Push Tokens And Multiple Profiles

Klaviyo SDK will disassociate the device push token from the current profile whenever it is reset by calling

> `set(profile:)` or `resetProfile`. You should call `set(pushToken:)` again after resetting the currently tracked profile

Reviewer 2: This is no longer true.

Reviewer 2: This should say the sdk will automatically associate the new token with the new profile.

Reviewer 3: Would it make sense to simplify this section to something like the following?

Whenever you call 'set(Profile)' or 'resetProfile', the KlaviyoSDK will automatically remove the token from the previous profile and move it to the new profile. This process prevents the device from receiving push notifications intended for the old profile.

Author: Based on the internal discussion we decided to remove this section and add the comments to the API. Seemed like initialize, resetProfile already had the comments. Added a comment to the set profile.

to explicitly associate the device token to the new profile.

### Request Push Notification Permission

Now that we have the push token, in order to send push notifications to your users, you must request permission to send push notifications.
Add the following code to your application wherever you would like to prompt users to register for push notifications.

> This is often included within [`application:didFinishLaunchingWithOptions:`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1622921-application)

Reviewer 2: Is this really a best practice...I think it's fine for now though.

Author: you mean don't suggest they request push permission during `didFinishLaunchingWithOptions`?

Reviewer 2: Yah I mean it would essentially prompt when the app starts...I wonder if you could get better opt-in rates with a different strategy.

Author: Thanks. I updated this section to include apple docs and made it clear that app start up isn't the best location for requesting push permission.

* Individual's phone number in [E.164 format](https://help.klaviyo.com/hc/en-us/articles/360046055671#h_01HE5ZYJEAHZKY6WZW7BAD36BG)

These above Identifiers are persisted to local storage so that the SDK can keep track of the current user/profile for you when you make event requests or want to set a push token etc.

Profile identifiers can be set all at once or individually. Either way, the SDK will group and batch API calls to improve performance.

The following code demonstrates how to set profile identifiers:

```swift
// organization, title, image, location and additional properties (dictionary) can also be set using the below constructor
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// or setting individual properties
KlaviyoSDK().set(profileAttribute: .firstName, value: "Blob")
KlaviyoSDK().set(profileAttribute: .lastName, value: "Jr.")
```

### Reset Profile
To start a new profile altogether (e.g. if a user logs out) either call `KlaviyoSDK().resetProfile()` to clear the currently tracked profile identifiers,
or use `KlaviyoSDK().set(profile: profile)` to overwrite it with a new profile object.

> <TODO: Should we talk about what happens to a push token here?>

Reviewer 2: Probalbly either here or in the section about push tokens.

Reviewer 4: Ya looks like this is actually already covered by the `Push Tokens And Multiple Profiles` section? Maybe can just remove this TODO?

Author: Yeah I added this TODO before adding that section. I'll remove the todo.

> <TODO: Should we talk about what happens to a push token here?>

Reviewer 3: I think that's a good idea. It would make sense to say that the token will move to the new profile.

Author: There is a section called `Push Tokens And Multiple Profiles` which address this in push notifications section. I added this todo before adding that in. Removed the todo now.

```swift
// start a profile for Blob Jr.
let profile = Profile(email: "junior@blob.com",  firstName: "Blob",  lastName: "Jr.")
KlaviyoSDK().set(profile: profile)

// stop tracking Blob Jr.
KlaviyoSDK().resetProfile()

// start a profile for Robin Hood

> let profile = Profile(email: "junior@blob.com",  firstName: "Robin",  lastName: "Hood")

Reviewer 2: maybe change the email here too?

Author: Are you saying RobinHood can't have junior@blob.com as an email id? JK, changed!

KlaviyoSDK().set(profile: profile)
```
### Anonymous Tracking Notice

> By default, Klaviyo will begin tracking unidentified users in your app when a push token is set or we receive an event to track.

Author: > By default, Klaviyo will begin tracking unidentified users in your app when a push token is set, or if it receives an event to track.
How does this sound?

> By default, Klaviyo will begin tracking unidentified users in your app when a push token is set or we receive an event to track.

Reviewer 2: Yah was thinking something like that!

> By default, Klaviyo will begin tracking unidentified users in your app when a push token is set or we receive an event to track.

Reviewer 3: If you do not identify an app user as outlined above, Klaviyo will automatically create an anonymous profile for unidentified users when you set a push token or track an event. Klaviyo identifies these profiles through an anonymous ID, a randomly generated UUID.

> By default, Klaviyo will begin tracking unidentified users in your app when a push token is set or we receive an event to track.

Author: Done.

> This means you will be able to track events or set a push notifications token from users in your app without any user information provided.

Reviewer 3: Using this ID, you will be able to track events or set a push token for an app user without any other user-identifying information provided.

> This means you will be able to track events or set a push notifications token from users in your app without any user information provided.

Reviewer 4: another grammar nit: getting a little sidetracked by the pluralization here. is it meant to be
`This means you will be able to track events or set push notification tokens from users in your app without any user information provided.`
or
`This means you will be able to track events or set a push notification token from users in your app without any user information provided.`

Essentially there's a trailing `s` on `push notifications` there

Author: yeah you are right. It's notification. Fixed.

When an email or other primary identifier is provided, Klaviyo will merge the data from the anonymous user to a new identified user.

Prior to version 1.7.0, the Klaviyo SDK used the [Apple identifier for vendor (IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) to facilitate anonymous tracking.

> Starting with version 1.7.0, the SDK will use a cached UUID that is generated when the SDK is initialized.

Author: Updated to - 

> Prior to version 1.7.0, the Klaviyo SDK used the [Apple identifier for vendor (IDFV)](https://developer.apple.com/documentation/uikit/uidevice/1620059-identifierforvendor) to facilitate anonymous tracking.
> Starting with version 1.7.0, the SDK will use a cached UUID.

For existing anonymous profiles using IDFV, the SDK will continue to use IDFV, instead of generating a new UUID.

## Event tracking

The SDK provides tools for tracking events that users perform on your app via the [Create Client Event API](https://developers.klaviyo.com/en/reference/create_client_event).
Below is an example of how to track an event:

```swift
// using a predefined event name
let event = Event(name: .StartedCheckoutMetric,
                  properties: [

>                         "product.1": "t-shirt.99",

Reviewer 2: These properties look a little odd but maybe they are derived from a real event?

Author: No I just came up with them. What would you suggest?

Reviewer 2: like name or color or size or something.

Reviewer 4: Ya pretty generic props should be fine here. Pretty sure what we put in the Android one was like that too so can just copy from that if you are unsure

Author: That makes more sense. Updated.

### Prerequisites

* An apple developer [account](https://developer.apple.com/).
* Configure [iOS push notifications](https://help.klaviyo.com/hc/en-us/articles/360023213971) in Klaviyo account settings.

### Setup

* Enable push notification capabilities in your Xcode project. The section "Enable the push notification capability" in this [Apple developer guide](https://developer.apple.com/documentation/usernotifications/registering_your_app_with_apns#2980170) provides detailed instructions.
* If you intend to use rich push notifications add a [Notification service extension](https://developer.apple.com/documentation/usernotifications/unnotificationserviceextension) to your xcode project.
A notification service app extension ships as a separate bundle inside your iOS app. To add this extension to your app:
  * Select File > New > Target in Xcode.
  * Select the Notification Service Extension target from the iOS > Application extension section.
  * Click Next.
  * Specify a name and other configuration details for your app extension.
  * Click Finish.

> ⚠️ By default the deployment target of your notification service extension might be the latest iOS version and not

> the minimum you want to support. This may cause push notifications to not show the attached media in devices whose

Reviewer 3: Should this be "on devices"?

iOS versions are lower than the deployment target of the notification service extension. ⚠️

### Collecting Push Token

In order to send push notifications to your users, you must collect their push tokens and register them with Klaviyo.

> This is done via the `KlaviyoSDK().set(pushToken:)` method, which registers push token and current authorization state

Reviewer 3: Should be "which registers a push token"

Author: Done.

You should see a file called `NotificationService.swift` under the notification service extension target (created during setup).
From here on depending on which dependency manager you use the steps would look slightly different:

<details>
<summary> Swift Package Manager(SPM) </summary>

- Tap on the newly created notification service extension target
  - Under General > Frameworks and libraries add `KlaviyoSwiftExtension` using the + button at the bottom left.
  - Then in the `NotificationService.swift` file add the code for the two required delegates from [this](Examples/KlaviyoSwiftExamples/SPMExample/NotificationServiceExtension/NotificationService.swift) file. This sample covers calling into Klaviyo so that we can download and attach the media to the push notification.
</details>

<details>
<summary> Cocoapods </summary>

- In your `Podfile` add in `KlaviyoSwiftExtension` as a dependency to the newly added notification service extension target.

  Example:

  ```
  target 'NotificationServiceExtension' do

>       pod 'KlaviyoSwiftExtension', '2.1.0-beta1'

Author: Removed the version so that it defaults to the latest. Good call out.

	    }

	    // Enable or disable features based on the authorization status.
	    // if you didn't register for remote notifications above you can call `registerForRemoteNotifications` here
	    // Klaviyo SDK will automatically update the authorization status on next app launch
	}

    return true
}
```

### Receiving Push Notifications

> #### Tracking Opens

Reviewer: ```suggestion
#### Tracking Open Events
```

Implement the [`userNotificationCenter:didReceive:withCompletionHandler`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649501-usernotificationcenter)

> and [`userNotificationCenter:willPresent:withCompletionHandler`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter) to handle push notifications received whe the app is in the background and foreground.

Reviewer 3: It looks like the "n" is missing from "received when"

Reviewer: ```suggestion
and [`userNotificationCenter:willPresent:withCompletionHandler`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter) to handle push notifications received when the app is in the background and foreground.
```

Author: I already made the change. Thanks, Evan.

> method in your app delegate to track when a user opens a push notification. This will handle tracking opens for when the app is backgrounded and the user taps on the notification.

Reviewer 4: Seems the start of this sentence is missing?

Author: Not really, I was just trying to get to the point directly but I did notice this read weird so changed it to include some context like - 
When a user taps on a push notification, Implement  [`userNotificationCenter:didReceive:withCompletionHandler`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649501-usernotificationcenter)
and [`userNotificationCenter:willPresent:withCompletionHandler`](https://developer.apple.com/documentation/usernotifications/unusernotificationcenterdelegate/1649518-usernotificationcenter) in your application delegate to handle receiving push notifications 
when the app is in the background and foreground respectively.

    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        didReceive response: UNNotificationResponse,
        withCompletionHandler completionHandler: @escaping () -> Void) {
        // If this notification is Klaviyo's notification we'll handle it
        // else pass it on to the next push notification service to which it may belong
        let handled = KlaviyoSDK().handle(notificationResponse: response, withCompletionHandler: completionHandler)
        if !handled {
            completionHandler()
        }
    }

    // below method is called when the app receives push notifications when the app is the foreground
    func userNotificationCenter(
        _ center: UNUserNotificationCenter,
        willPresent notification: UNNotification,
        withCompletionHandler completionHandler: @escaping (UNNotificationPresentationOptions) -> Void) {
        if #available(iOS 14.0, *) {
            completionHandler([.list, .banner])
        } else {
            completionHandler([.alert])
        }
    }
}
```
Once your first push notifications are sent and opened, you should start to see _Opened Push_ metrics within your Klaviyo dashboard.

> #### Rich push notifications

Reviewer: ```suggestion
#### Rich Push
```

>  ℹ️ Rich push notifications are supported in SDK version [2.2.0](https://github.com/klaviyo/klaviyo-swift-sdk/releases/tag/2.2.0) and higher

> Rich push notification is the ability to add images to your push notification messages that Apple has supported since iOS 10.

Reviewer 2: Reads funny to me a little...
Maybe Rich push notifications are push notifications that include images. Apple has supported this feature since iOS 10.

Reviewer: ```suggestion
[Rich Push](https://help.klaviyo.com/hc/en-us/articles/16917302437275) is the ability to add images to
push notification messages.
```
is the opening line i've used in the other 2 readmes

  - Then in the `NotificationService.swift` file add the code for the two required delegates from [this](Examples/KlaviyoSwiftExamples/SPMExample/NotificationServiceExtension/NotificationService.swift) file. This sample covers calling into Klaviyo so that we can download and attach the media to the push notification.
</details>

<details>
<summary> Cocoapods </summary>

- In your `Podfile` add in `KlaviyoSwiftExtension` as a dependency to the newly added notification service extension target.

  Example:

  ```
  target 'NotificationServiceExtension' do
      pod 'KlaviyoSwiftExtension', '2.1.0-beta1'
  end
  ```

  Be sure to replace the name of your notification service extension target above.

  - Once you've added in the dependency make sure to `pod install`.
  - Then in the `NotificationService.swift` file add the code for the two required delegates from [this](Examples/KlaviyoSwiftExamples/CocoapodsExample/NotificationServiceExtension/NotificationService.swift) file. This sample covers calling into Klaviyo so that we can download and attach the media to the push notification.
</details>

> * Test your rich push notifications

Reviewer 3: It feels like we're missing a sentence or two after this. Are the next lines bullets for how someone could test? Should we say something like "To test rich push notifications, you will need three things:"

Author: added

>   * Any push notifications tester like apple official [push notification console](https://developer.apple.com/notifications/push-notifications-console/) or a third party software such as [this](https://github.com/onmyway133/PushNotifications).

Reviewer 4: nit: `like Apple's official`

Just capitalizing Apple and adding `'s`

  * A push notification payload that resembles what Klaviyo would send to you. The below payload should work as long as the image is valid:

    ```json
    {
    "aps": {
    "alert": {
      "title": "Free apple vision pro",
      "body": "Free Apple vision pro when you buy a Klaviyo subscription."
    },
    "mutable-content": 1
    },
    "rich-media": "https://www.apple.com/v/apple-vision-pro/a/images/overview/hero/portrait_base__bwsgtdddcl7m_large.jpg",
    "rich-media-type": "jpg"
    }
    ```
  * A real device's push notification token. This can be printed out to the console from the `didRegisterForRemoteNotificationsWithDeviceToken` method in `AppDelegate`.

>   * Once we have these three things we can then use the push notifications tester and send a local push notification to make sure that everything was set up correctly.

Reviewer 3: It feels like we should change "we" to "you." "Once you have these three things, you can then use the push notifications tester and send a local push notification to make sure that everything was set up correctly."

Author: Done.

#### Deep Linking

>  ℹ️  Your app needs to use version 1.7.2 at a minimum in order for the below steps to work.

> 

Reviewer 3: I feel like we need a description of what deep linking is just like we do in the Android readme. 

[Deep Links](https://help.klaviyo.com/hc/en-us/articles/14750403974043) allow you to navigate to a particular page within your app in response to the user opening a notification.

Author: Done.

> 

Reviewer 3: I would remove lines 376 to 379 above. I don't think those are necessary if we add the description above.

Author: Done.

> In order for deep linking to work, there are a few configurations that are needed and these are no different from what are required for handling deep linking in general and [Apple documentation](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app) on this can be followed in conjunction with the steps highlighted here:

Reviewer 3: This wording feels awkward. How about this?

You need to configure deep links in your app for them to work. The configuration process for Klaviyo is no different than what is required for handling deep linking in general, so you can follow the [Apple documentation](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app) for deep linking in conjunction with the steps outlined here:

Author: Thanks for the suggestion. Done.

> ##### Option 1: URI Schemes

Reviewer 3: It looks like we use URL scheme everywhere else. Should it be URI or URL?

Reviewer 4: Typically I've seen them called URI schemes but the name is interchangeable I guess. Agree maybe consistency would be clearest though.

Author: Going to switch to using URL since that's what Apple seems to use in their [docs](https://developer.apple.com/documentation/xcode/defining-a-custom-url-scheme-for-your-app).

Author: Done.

> ##### Option 1: URI Schemes

Reviewer 4: One thing I commented on the Android README when Evan was doing the redux over that side is that the title nesting is getting a little crazy at this level. I wonder if we should just remove anything around universal links and just talk about URL scheme setup steps so that we dont need to get 5-6 layers deep.

Cant really remember where we landed with universal links stuff and having that in the readme though. Maybe @mark-piana would be clearer on that front

Reviewer 3: I feel like it's good information to have in the readme if we support it. I know we're not hosting anything on our side, but I think we'll get questions if we remove information about universal links, so my vote would be to leave it in.

> If you do not need universal link support you can instead implement url schemes for your app and the deepLinkHandler as indicated in Option 1 can be omitted. The Klaviyo SDK will follow all url automatically in this case.

Reviewer 3: I think we can remove this line since it references the old order of the readme.

Reviewer 3: Instead maybe we should say something like "URL schemes are the traditional and simpler way of deep linking from a push notification to your app. However, these links will only work if your mobile app is installed on a device and will not be understood by a web browser if, for example, you want to link from an email to your app."

Author: Hmm, so yeah I missed updating this but I think this is calling out that if your code is not going to use universal links that you don't need to implement some code related to universal links. But now that we switched the ordering, we can remove that and add context with what you suggested here.

</array>
```

###### Step 3: Implement handling deep links in your app

> Steps 1 & 2 set your app up for receiving deep links but now is when you need to figure out how to handle them within your app.

Reviewer 3: This wording feels strange to me. It feels like we're saying "you need to figure out how to handle deep links, but we're not going to tell you. Good luck!"

How about "Steps 1 and 2 enable your app to receive deep links, but you also need to handle these links within your app. Follow the relevant steps below depending on whether you are using UIKit or SwiftUI."

Author: LOL does sound that way! Took your suggestion but made a slight change to avoid some redundancy.

###### Step 3: Implement handling deep links in your app

Steps 1 and 2 enable your app to receive deep links, but you also need to handle these links within your app. 
This is done by implementing the [`application:openURL:options:`](https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623112-application) 
method in your app delegate.

Example:

```swift
func application(
    _ app: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey : Any] = [:]
) -> Bool {
    guard let components = NSURLComponents(url: url, resolvingAgainstBaseURL: true)
    else {
       print("Invalid deep linking URL")
       return false
    }

    print("components: \(components.debugDescription)")

    return true
}
```

If you are using SwiftUI, then you can implement [`onOpenURL(perform:)`](<https://developer.apple.com/documentation/swiftui/view/onopenurl(perform:)>) as a view modifier in the view you intent to handle deep links. This may or may not be the root of your scene.

Example:

```swift
@main
struct MyApplication: App {
  var body: some Scene {
    WindowGroup {
      ContentView()
        .onOpenURL { url in
          // handle the URL that must be opened
        }
    }
  }
}
```

Finally, we have an example app (`Examples/KlaviyoSwiftExamples`) in the SDK repo that you can reference to get an example of how to implement deep links in your app.

Once the above steps are complete, you can send push notifications from the Klaviyo Push editor within the Klaviyo website. 
Here you can build and send a push notification through Klaviyo to make sure that the URL shows up in the handler you implemented in Step 3.

Additionally, you can also locally trigger a deep link to make sure your code is working using the below command in the terminal.

`xcrun simctl openurl booted {your_URL_here}`

`xcrun simctl openurl booted {your_URL_here}`

##### Option 2: Universal links

[Universal links](https://developer.apple.com/ios/universal-links/) are a more modern way of handling deep links and are recommended by Apple.
They are more secure and provide a better user experience. However, unlike URL schemes they require a bit more setup that is highlighted in [these](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppSearch/UniversalLinks.html) Apple docs.

Once you have the setup from the Apple docs in place you will need to modify the push open tracking as described below:

```swift
extension AppDelegate: UNUserNotificationCenterDelegate {
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        let handled = KlaviyoSDK().handle(notificationResponse: response, withCompletionHandler: completionHandler) { url in
            print("deep link is ", url)
        }
        if !handled {
           // not a klaviyo notification should be handled by other app code
        }
    }
}
```

> Note that the deep link handler will be called back on the main thread. If you want to handle URL schemes in addition to universal links you implement them as described in [Option1: URL Schemes](#option-1-URL-schemes).

Reviewer: ```suggestion
Note that the deep link handler will be called back on the main thread. If you want to handle URL schemes in addition to universal links you implement them as described in [Option 1: URL Schemes](#option-1-URL-schemes).
```

`xcrun simctl openurl booted {your_URL_here}`

##### Option 2: Universal links

[Universal links](https://developer.apple.com/ios/universal-links/) are a more modern way of handling deep links and are recommended by Apple.

> They are more secure and provide a better user experience. However, unlike URI schemes they require a bit more setup that is highlighted in [these](https://developer.apple.com/library/archive/documentation/General/Conceptual/AppSearch/UniversalLinks.html) apple docs.

Reviewer 3: Apple should be capitalized

Author: Done.

> Once you have the setup from the apple docs in place you will need to modify the push open tracking as described below:

Reviewer 3: Apple should be capitalized

Author: Done.

```swift
extension AppDelegate: UNUserNotificationCenterDelegate {
    func userNotificationCenter(_ center: UNUserNotificationCenter, didReceive response: UNNotificationResponse, withCompletionHandler completionHandler: @escaping () -> Void) {
        let handled = KlaviyoSDK().handle(notificationResponse: response, withCompletionHandler: completionHandler) { url in
            print("deep link is ", url)
        }
        if !handled {
           // not a klaviyo notification should be handled by other app code
        }
    }
}
```

Note that the deep link handler will be called back on the main thread. If you want to handle URI schemes in addition to universal links you implement them as described in [Option1: URI Schemes](#option-1-uri-schemes).

## Additional Details

### Sandbox Support

> Apple has two environments with push notification support - production and Sandbox.

Reviewer 3: Production should be capitalized to be consistent

Author: Done.

The Production environment supports sending push notifications to real users when an app is published in the App Store or TestFlight.
In contrast, Sandbox applications that support push notifications are those signed with iOS Development Certificates, instead of iOS Distribution Certificates.

> Sandbox acts as a staging environment, allowing you to test your applications in a environment similar to but distinct from production without having to worry about sending messages to real users.

Reviewer 3: Production should be capitalized

Author: Done.

> Our SDK supports the use of Sandbox for push as well.

Reviewer 3: We should specify that this is true for v2.2.0 and higher. I think that's the right version, but I'm not 100% sure.

Author: Thanks Mark. Updated to add a info block - 
> ℹ️ Sandbox support is available in SDK version 2.2.0 and higher

| Network   | Interval   |
| :-------- | :--------- |
| WWAN/Wifi | 10 seconds |
| Cellular  | 30 seconds |

Connection determination is based on notifications from our reachability service.
When there is no network available, the SDK will cache data until the network becomes available again.
All data sent by the SDK should be available shortly after it is flushed by the SDK.

### Retries
The SDK will retry API requests that fail under certain conditions. For example, if a network timeout occurs, the request will be retried on the next flush interval.
In addition, if the SDK receives a rate limiting error `429` from the Klaviyo API, it will use exponential backoff with jitter to retry the next request.

### License
KlaviyoSwift is available under the MIT license. See the LICENSE file for more info.

### UserDefaults access (SDK version < 3.0.0)
As of fall 2023, Apple requires apps that access specific Apple APIs to provide a reason for this access.
Previous versions of the Klaviyo SDK used UserDefaults to store data about the current user.

> Today, when the SDK starts up, it must access this data to migrate to a new format.

Reviewer 2: Should we change `today` to say in versions below 3.0.0 or something...

Author: Hmm good call. Can we get rid of this section now that we don't use user defaults? Yeah some of our older versions do but we've never heard anyone ask about this so far. CC: @mark-piana

Reviewer 3: You're probably right. I'm not sure why we'd need to include this anymore since anyone just starting to implement things wouldn't care about this information.

Reviewer: Yeah IDK how important this whole section is, maybe it could be called out in the migration guide or something, but doesn't feel like its relevant in the README for v3.0+

Author: I'll just remove it for now. I don't think Apple is strict about enforcing this either.
